File: experimental-plugins/agentic-workflows/src/agentic_workflows/ralph_loop.py
"""Ralph Wiggum loop state management and completion detection.

This module implements the Ralph Wiggum iterative loop pattern for Claude workflows.
Each iteration creates a fresh Claude session with the same prompt.
Completion is signaled via structured JSON output from Claude.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import yaml


@dataclass
class RalphLoopState:
    """State for a Ralph Wiggum loop execution."""

    active: bool
    iteration: int
    max_iterations: int
    completion_promise: str
    started_at: str
    prompt: str


def get_ralph_state_path(workflow_id: str, step_name: str, repo_root: Path) -> Path:
    """Get path to ralph loop state file."""
    safe_name = re.sub(r"[^\w\-]", "_", step_name)
    return repo_root / "agentic" / "workflows" / workflow_id / f"ralph-{safe_name}.md"


def create_ralph_state(
    workflow_id: str,
    step_name: str,
    prompt: str,
    max_iterations: int,
    completion_promise: str,
    repo_root: Path,
) -> RalphLoopState:
    """Create initial ralph loop state file."""
    state = RalphLoopState(
        active=True,
        iteration=1,
        max_iterations=max_iterations,
        completion_promise=completion_promise,
        started_at=datetime.now(timezone.utc).isoformat(),
        prompt=prompt,
    )

    _save_ralph_state(workflow_id, step_name, state, repo_root)
    return state


def load_ralph_state(
    workflow_id: str, step_name: str, repo_root: Path
) -> RalphLoopState | None:
    """Load ralph loop state from file."""
    state_path = get_ralph_state_path(workflow_id, step_name, repo_root)
    if not state_path.exists():
        return None

    content = state_path.read_text(encoding="utf-8")
    return _parse_ralph_state(content)


def update_ralph_iteration(
    workflow_id: str, step_name: str, repo_root: Path
) -> RalphLoopState | None:
    """Increment iteration counter in ralph state."""
    state = load_ralph_state(workflow_id, step_name, repo_root)
    if not state:
        return None

    state.iteration += 1
    _save_ralph_state(workflow_id, step_name, state, repo_root)
    return state


def deactivate_ralph_state(
    workflow_id: str, step_name: str, repo_root: Path
) -> None:
    """Mark ralph loop as inactive (completed or stopped)."""
    state = load_ralph_state(workflow_id, step_name, repo_root)
    if state:
        state.active = False
        _save_ralph_state(workflow_id, step_name, state, repo_root)


def delete_ralph_state(workflow_id: str, step_name: str, repo_root: Path) -> None:
    """Delete ralph loop state file."""
    state_path = get_ralph_state_path(workflow_id, step_name, repo_root)
    if state_path.exists():
        state_path.unlink()


def _save_ralph_state(
    workflow_id: str, step_name: str, state: RalphLoopState, repo_root: Path
) -> None:
    """Save ralph loop state to markdown file with YAML frontmatter."""
    state_path = get_ralph_state_path(workflow_id, step_name, repo_root)
    state_path.parent.mkdir(parents=True, exist_ok=True)

    frontmatter = {
        "active": state.active,
        "iteration": state.iteration,
        "max_iterations": state.max_iterations,
        "completion_promise": state.completion_promise,
        "started_at": state.started_at,
    }

    content = f"""---
{yaml.dump(frontmatter, default_flow_style=False).strip()}
---

# Ralph Loop Prompt

{state.prompt}
"""

    state_path.write_text(content, encoding="utf-8")


def _parse_ralph_state(content: str) -> RalphLoopState | None:
    """Parse ralph state from markdown with YAML frontmatter."""
    if not content.startswith("---"):
        return None

    parts = content.split("---", 2)
    if len(parts) < 3:
        return None

    try:
        frontmatter = yaml.safe_load(parts[1])
    except yaml.YAMLError:
        return None

    prompt_section = parts[2].strip()
    prompt = ""
    if "# Ralph Loop Prompt" in prompt_section:
        prompt = prompt_section.split("# Ralph Loop Prompt", 1)[1].strip()
    else:
        prompt = prompt_section

    return RalphLoopState(
        active=frontmatter.get("active", False),
        iteration=frontmatter.get("iteration", 1),
        max_iterations=frontmatter.get("max_iterations", 5),
        completion_promise=frontmatter.get("completion_promise", "COMPLETE"),
        started_at=frontmatter.get("started_at", ""),
        prompt=prompt,
    )


@dataclass
class CompletionResult:
    """Result of completion promise detection."""

    is_complete: bool
    promise_matched: bool
    promise_value: str | None
    output: str


def detect_completion_promise(
    output: str, expected_promise: str
) -> CompletionResult:
    """Detect completion promise in Claude's JSON output.

    Claude should output a JSON block with the following structure:
    ```json
    {
      "ralph_complete": true,
      "promise": "COMPLETE"
    }
    ```

    Args:
        output: The raw output from Claude
        expected_promise: The expected promise text

    Returns:
        CompletionResult indicating whether the promise was matched
    """
    # Try to find JSON in the output
    json_patterns = [
        # Code block with json tag
        r"```json\s*\n?(.*?)\n?```",
        # Code block without tag
        r"```\s*\n?(\{.*?\})\n?```",
        # Raw JSON object (greedy match for nested objects)
        r"(\{[^{}]*\"ralph_complete\"[^{}]*\})",
    ]

    for pattern in json_patterns:
        matches = re.findall(pattern, output, re.DOTALL | re.IGNORECASE)
        for match in matches:
            try:
                data = json.loads(match.strip())
                if isinstance(data, dict):
                    ralph_complete = data.get("ralph_complete", False)
                    promise_value = data.get("promise", "")

                    if ralph_complete:
                        promise_matched = (
                            str(promise_value).strip().upper()
                            == str(expected_promise).strip().upper()
                        )
                        return CompletionResult(
                            is_complete=True,
                            promise_matched=promise_matched,
                            promise_value=str(promise_value),
                            output=output,
                        )
            except json.JSONDecodeError:
                continue

    return CompletionResult(
        is_complete=False,
        promise_matched=False,
        promise_value=None,
        output=output,
    )


def build_ralph_system_message(
    iteration: int, max_iterations: int, completion_promise: str
) -> str:
    """Build system message for Ralph loop iteration.

    This message is prepended to the prompt to inform Claude about the loop context.
    """
    return f"""## Ralph Wiggum Loop - Iteration {iteration}/{max_iterations}

You are in a Ralph Wiggum iterative loop. Each iteration starts a fresh session.

**Completion Signal:**
When your task is FULLY complete, output a JSON block:
```json
{{
  "ralph_complete": true,
  "promise": "{completion_promise}"
}}
```

**IMPORTANT:**
- Only output the completion JSON when the task is genuinely finished
- Do not output false promises to exit early
- If you cannot complete the task, continue working - you have {max_iterations - iteration} iterations remaining

---

"""

File: src/ralph-loop.ts
/** Ralph Wiggum loop state management and completion detection. */

import { existsSync, mkdirSync, readFileSync, unlinkSync, writeFileSync } from "node:fs";
import path from "node:path";
import yaml from "js-yaml";

// --- Types ---

export interface RalphLoopState {
	active: boolean;
	iteration: number;
	maxIterations: number;
	completionPromise: string;
	startedAt: string;
	prompt: string;
}

export interface CompletionResult {
	isComplete: boolean;
	promiseMatched: boolean;
	promiseValue: string | null;
	output: string;
	isFailed: boolean;
	failureReason: string | null;
}

// --- State path ---

export function getRalphStatePath(workflowId: string, stepName: string, outputDir: string): string {
	const safeName = stepName.replace(/[^\w-]/g, "_");
	return path.join(outputDir, `ralph-${safeName}.md`);
}

// --- State management ---

export function createRalphState(
	workflowId: string,
	stepName: string,
	prompt: string,
	maxIterations: number,
	completionPromise: string,
	outputDir: string,
): RalphLoopState {
	const state: RalphLoopState = {
		active: true,
		iteration: 1,
		maxIterations,
		completionPromise,
		startedAt: new Date().toISOString(),
		prompt,
	};

	saveRalphState(workflowId, stepName, state, outputDir);
	return state;
}

export function loadRalphState(
	workflowId: string,
	stepName: string,
	outputDir: string,
): RalphLoopState | null {
	const statePath = getRalphStatePath(workflowId, stepName, outputDir);
	if (!existsSync(statePath)) return null;

	const content = readFileSync(statePath, "utf-8");
	return parseRalphState(content);
}

export function updateRalphIteration(
	workflowId: string,
	stepName: string,
	outputDir: string,
): RalphLoopState | null {
	const state = loadRalphState(workflowId, stepName, outputDir);
	if (!state) return null;

	state.iteration += 1;
	saveRalphState(workflowId, stepName, state, outputDir);
	return state;
}

export function deactivateRalphState(
	workflowId: string,
	stepName: string,
	outputDir: string,
): void {
	const state = loadRalphState(workflowId, stepName, outputDir);
	if (state) {
		state.active = false;
		saveRalphState(workflowId, stepName, state, outputDir);
	}
}

export function deleteRalphState(workflowId: string, stepName: string, outputDir: string): void {
	const statePath = getRalphStatePath(workflowId, stepName, outputDir);
	if (existsSync(statePath)) {
		unlinkSync(statePath);
	}
}

// --- Internal helpers ---

function saveRalphState(
	workflowId: string,
	stepName: string,
	state: RalphLoopState,
	outputDir: string,
): void {
	const statePath = getRalphStatePath(workflowId, stepName, outputDir);
	const dir = path.dirname(statePath);
	mkdirSync(dir, { recursive: true });

	const frontmatter = {
		active: state.active,
		iteration: state.iteration,
		max_iterations: state.maxIterations,
		completion_promise: state.completionPromise,
		started_at: state.startedAt,
	};

	const yamlStr = yaml.dump(frontmatter, { flowLevel: -1 }).trim();
	const content = `---
${yamlStr}
---

# Ralph Loop Prompt

${state.prompt}
`;

	writeFileSync(statePath, content, "utf-8");
}

export function parseRalphState(content: string): RalphLoopState | null {
	if (!content.startsWith("---")) return null;

	const parts = content.split("---");
	if (parts.length < 3) return null;

	let frontmatter: Record<string, unknown>;
	try {
		frontmatter = yaml.load(parts[1]) as Record<string, unknown>;
	} catch {
		return null;
	}

	const promptSection = parts.slice(2).join("---").trim();
	let prompt: string;
	if (promptSection.includes("# Ralph Loop Prompt")) {
		prompt = promptSection.split("# Ralph Loop Prompt")[1].trim();
	} else {
		prompt = promptSection;
	}

	return {
		active: (frontmatter.active as boolean) ?? false,
		iteration: (frontmatter.iteration as number) ?? 1,
		maxIterations: (frontmatter.max_iterations as number) ?? 5,
		completionPromise: (frontmatter.completion_promise as string) ?? "COMPLETE",
		startedAt: (frontmatter.started_at as string) ?? "",
		prompt,
	};
}

// --- Completion detection ---

export function detectCompletionPromise(output: string, expectedPromise: string): CompletionResult {
	const jsonPatterns = [
		/```json\s*\n?(.*?)\n?```/gis,
		/```\s*\n?(\{.*?\})\n?```/gis,
		/(\{[^{}]*"ralph_complete"[^{}]*\})/gi,
		/(\{[^{}]*"ralph_failed"[^{}]*\})/gi,
	];

	for (const pattern of jsonPatterns) {
		pattern.lastIndex = 0;
		const matches = [...output.matchAll(pattern)];
		for (const match of matches) {
			try {
				const data = JSON.parse(match[1].trim());
				if (typeof data !== "object" || data === null) continue;

				// Check for completion signal
				if (data.ralph_complete) {
					const promiseValue = String(data.promise ?? "");
					const promiseMatched =
						promiseValue.trim().toUpperCase() === expectedPromise.trim().toUpperCase();
					return {
						isComplete: true,
						promiseMatched,
						promiseValue,
						output,
						isFailed: false,
						failureReason: null,
					};
				}

				// Check for failure signal
				if (data.ralph_failed) {
					const failureReason = String(data.reason ?? "Unknown error");
					return {
						isComplete: true,
						promiseMatched: false,
						promiseValue: null,
						output,
						isFailed: true,
						failureReason,
					};
				}
			} catch {
				// Invalid JSON, skip
			}
		}
	}

	return {
		isComplete: false,
		promiseMatched: false,
		promiseValue: null,
		output,
		isFailed: false,
		failureReason: null,
	};
}

// --- System message builder ---

export function buildRalphSystemMessage(
	iteration: number,
	maxIterations: number,
	completionPromise: string,
): string {
	return `## Ralph Wiggum Loop - Iteration ${iteration}/${maxIterations}

You are in a Ralph Wiggum iterative loop. Each iteration starts a fresh session.

**Completion Signal:**
When your task is FULLY complete, output a JSON block:
\`\`\`json
{
  "ralph_complete": true,
  "promise": "${completionPromise}"
}
\`\`\`

**Failure Signal:**
Only output this after exhausting ALL alternatives (see retry rules below):
\`\`\`json
{
  "ralph_failed": true,
  "reason": "<brief description of the blocking error and all approaches attempted>"
}
\`\`\`

**Tool Retry Rules:**
If a tool call is denied or fails, you MUST try alternative approaches before giving up. Attempt at least 3 different strategies. For example, if editing a file fails:
1. Try the Write tool to overwrite the file with updated content
2. Try Bash with a shell command (e.g., echo/printf with redirection)
3. Try Bash with sed or cat with heredoc to rewrite the file
Only output the failure JSON after ALL alternatives have been attempted and failed.

**IMPORTANT:**
- Only output the completion JSON when the task is genuinely finished
- Do not output false promises to exit early
- NEVER output the failure JSON after a single tool denial - always try alternative tools first
- If the task is incomplete but can still progress, continue working - you have ${maxIterations - iteration} iterations remaining

---

`;
}
